### ISSUE TITLE
[Feat]: add `anomaly_detection.anomaly_rate_by_/user/app` chart

### ISSUE THREAD

**Issue by andrewm4894** [netdata/netdata#14788]
### Problem

It would be useful if we could easily go "one step" below the existing `anomaly_detection.anomaly_rate` chart to easily see anomaly rates by each area of a node.

The idea here is to create a slightly lower level chart that shows anomaly rates by some grouping that provides some more granularity.

### Description

An `anomaly_detection.anomaly_rate_by_type` chart that has a line per type from the `api/v1/charts` response. 

https://london.my-netdata.io/api/v1/charts

This typically would be about 50 dimensions and so is a manageable number in terms of overhead and visual display on a single chart. 

More info on what "type" means can be found here: https://learn.netdata.cloud/docs/data-collection/chart-dimensions-contexts-and-families#context

> The apps portion of the context is the type, whereas anything after the . is specified either by the chart's developer or by the [family](https://learn.netdata.cloud/docs/data-collection/chart-dimensions-contexts-and-families#family).

> By default, a chart's type affects where it fits in the menu, while its family creates submenus.

So since "type" is a core part of how Netdata groups and relates charts, producing an anomaly rate per type will be a useful way to summarize anomaly rates at a level one step lower (and so more useful potentially) then the existing single line on `anomaly_detection.anomaly_rate` chart.

Here is a real example where its easy to just see that the spike in node anomaly rate is mainly in a cgroup where i triggered the anomaly

![Image](https://user-images.githubusercontent.com/2178292/226752497-02478ea7-cebd-4b9a-ba12-14e872d4b45e.png)

Here is a little clearer visualization to illustrate things: 

<img width="856" alt="image" src="https://user-images.githubusercontent.com/2178292/226753049-d27f7fce-49b8-4327-84e2-63a537b08e76.png">

Here is a little video showing why i think this will be useful:

https://user-images.githubusercontent.com/2178292/226881128-ff90bbdc-9455-4e1f-8209-d543597f5ea9.mp4

link to colab notebook that you can paste an agent dashboard url into and play with it yourself: https://colab.research.google.com/drive/1XNq_GE4rqey4biKR-I8nzOpeH9CXrLwD?usp=sharing

### Importance

really want

### Value proposition

1. an easy way to quickly see from one chart "where" a spike in anomaly rate may mostly be coming from.
2.
3.
...

### Proposed implementation

As part of the native ML implementation in the agent we should also produce a `anomaly_detection.anomaly_rate_by_type` chart.

**Comment by andrewm4894** [comment]
@vkalintiris how feasible do you think something like this would be to implement?

Just an `anomaly_detection.anomaly_rate_per_type` chart where each dim is the anomaly rate across all dims under each `type`.

**Comment by vkalintiris** [comment]
> @vkalintiris how feasible do you think something like this would be to implement?

Don't think it's difficult to implement. However, I'm of the opinion that dimensions/charts that are not created from "primary/raw" collectors should be generated (somehow) on the fly, in order to minimize the CPU/RAM/disk impact of the agent.

**Comment by andrewm4894** [comment]
So maybe we need some concept of a "virtual" or "dynamic" chart in the agent that is really just a view or qry over the db - @ktsaou  - thoughts?

**Comment by andrewm4894** [comment]
@vkalintiris @ktsaou I'd like us to do this PR to add this chart and refactor it out to be a dynamic or virtual chart later if we decide to. 

I think this chart merits being storage as a normal chart so that it can be used by health engine and is always available just like any other chart.

If we wanted something like `anomaly_detection.anomaly_rate_by_context` as an additional chart then for sure that would need to be done like a virtual chart but i think `anomaly_detection.anomaly_rate_by_type` is small and useful enough to just do natively as a sort of "quick win" for now.

**Comment by andrewm4894** [comment]
@ktsaou @vkalintiris it would also be great to be able to do an anomaly rate by app and user as a sort of synthetic metric like this. So building on to the more general case of #15102

**Comment by andrewm4894** [comment]
I have been discussing this with @vkalintiris 

example showing what i mean (colab notebook)[https://colab.research.google.com/drive/1dPqnTEUl5ZRLTp9rMwfuLMv07Rsr7th7?usp=sharing]

```
apps_dims_map['httpd']
```

so the `httpd` dim on the `anomaly_detection.anomaly_rate_by_app` chart would be the average of anomaly bit over all these dims:

aka - the anomaly rate for the `httpd` app according to netdata. similar we would do for each dim found anywhere under `users.*` charts.

```
['apps.cpu|httpd',
 'apps.mem|httpd',
 'apps.rss|httpd',
 'apps.vmem|httpd',
 'apps.threads|httpd',
 'apps.processes|httpd',
 'apps.uptime|httpd',
 'apps.cpu_user|httpd',
 'apps.cpu_system|httpd',
 'apps.voluntary_ctxt_switches|httpd',
 'apps.involuntary_ctxt_switches|httpd',
 'apps.swap|httpd',
 'apps.major_faults|httpd',
 'apps.minor_faults|httpd',
 'apps.preads|httpd',
 'apps.pwrites|httpd',
 'apps.lreads|httpd',
 'apps.lwrites|httpd',
 'apps.fds_open_limit|httpd',
 'apps.fds_open|httpd',
 'apps.fds_files|httpd',
 'apps.fds_sockets|httpd',
 'apps.fds_pipes|httpd',
 'apps.fds_inotifies|httpd',
 'apps.fds_eventfds|httpd',
 'apps.fds_timerfds|httpd',
 'apps.fds_signalfds|httpd',
 'apps.fds_eventpolls|httpd',
 'apps.fds_other|httpd']
```

**Comment by vkalintiris** [comment]
@andrewm4894 I don't see how we can implement this without hurting performance, because the agent's internal data structures are built around the hosts -> charts -> dimensions hierarchy. If we want to implement this in a performant way, we'd need to reorganize the way we are bookkeeping apps in the agent. FWIW, implementing the same functionality by chart `type` is both straight forward and performant.

Please, discuss this with @ktsaou (add me, if you need my input) and let me know how we should proceed.

**Comment by andrewm4894** [comment]
ok i see - then maybe it's looks like might be more of a good first use case for #15102 perhaps. 

in hindsight "...by_type" is probably confusing for users and "...by_app" and "...by_user" would be more intuitive and probably useful on balance.

**Comment by vkalintiris** [comment]
> ok i see - then maybe it's looks like might be more of a good first use case for https://github.com/netdata/netdata/issues/15102 perhaps.

No, it won't be. We can't calculate the anomaly rate per _app_ in an performant way because the agent itself does not have any notion of the _app_ concept (in layman's terms). We'd either have to introduce the concept of an _app_ to the agent (probably, by feeding this information at chart creation time, through the plugins), or improvise and do the bookkeeping ourselves which would not be performant.

**Comment by andrewm4894** [comment]
> We can't calculate the anomaly rate per app in an performant way because the agent itself does not have any notion of the app concept (in layman's terms)

i mean that this is probably the case for any random subset of metrics though right? 

synthetic metrics are about deriving a new virtual metric from some arbitrary subset of metrics in netdata - so in that case i'm thinking of this as an instance of a synthetic metric.

that said, if better or more straight forward to just treat this whole ask as a separate specific requirement just focused on what we need here then i'm down with that too.

**Comment by vkalintiris** [comment]
> i mean that this is probably the case for any random subset of metrics though right

Correct, my point was that for this specific use case synthetics metrics won't cut it, ie. we'd need to modify the agent to _understand_ apps, ie. organize things such as charts/dimensions around apps.

**Comment by andrewm4894** [comment]
> modify the agent to understand apps

cool, i think that might end up being a useful thing that could serve more than just this use case even (even though this use case is all i care about right now)

**Comment by andrewm4894** [comment]
@vkalintiris @ktsaou I have thought on this a bit and am thinking....

**Step 1 - By Type**

create `anomaly_detection.anomaly_rate_by_type` as @vkalintiris suggested is reasonably feasible. This will be useful as its more related to "physical" anomalies on the node.

**Step 2 - By App**

create `anomaly_detection.anomaly_rate_by_app` and do it properly. This will be useful as its more related to "what" (in terms of software) has anomalies on the node.

**Step 3 - By User**

create `anomaly_detection.anomaly_rate_by_user` and do it similar to step 2 learnings. This will be useful as its more related to "who" (in terms of software) has anomalies on the node.


End results being that having `anomaly_detection.anomaly_rate_by_type|app|user` actually covers 3 main ways you might want to quickly spot anomalies on your nodes - it is something more physical like disks or cpu etc, are there particular apps or users involved.

So i would propose that @vkalintiris start with step 1 as something we can get sooner (and i can help test and validate etc) and then following that we plan how to tackle step 2 properly.

**Comment by vkalintiris** [comment]
> So i would propose that @vkalintiris start with step 1 as something we can get sooner

Will start working on this and have a PR ready early next week.

**Comment by ilyam8** [comment]
Looks like the issue is stale. Feel free to open it again if anyone plans to work on it.
